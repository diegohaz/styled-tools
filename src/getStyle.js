// @flow
type InterpolatedFn = (props: Object) => any;
type Part = string | InterpolatedFn | Part[];

const mapRules = (props: Object) => (part: Part) => {
  if (typeof part === "function") {
    return mapRules(props)(part(props));
  }
  if (Array.isArray(part)) {
    return part.map(mapRules(props)).join("");
  }
  return part;
};

/**
 * Get the style rules from a given styled component.
 *
 * @example
 * import styled from "styled-components";
 * import { getStyle, prop } from "styled-tools";
 *
 * const Background = styled.div`
 *   display: block;
 *   background-color: ${prop("bg")};
 * `;
 *
 * const BlackBackground = styled.div`
 *   ${getStyle(Background, { bg: "black" })};
 * `;
 */
const getStyle = (
  styledComponent: Object = {},
  props: Object = styledComponent.defaultProps || {}
) => {
  if (!styledComponent.componentStyle) {
    throw new Error(
      "[styled-tools] getStyle expects to receive a styled component."
    );
  }
  return styledComponent.componentStyle.rules.map(mapRules(props)).join("");
};

export default getStyle;
