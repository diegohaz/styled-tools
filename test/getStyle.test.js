import styled, { css } from "styled-components";
import getStyle from "../src/getStyle";

test("no props", () => {
  const Comp = styled.div`
    display: block;
    margin: 10px 20px;
  `;
  expect(getStyle(Comp)).toMatchSnapshot();
});

test("with props", () => {
  const Comp = styled.div`
    display: block;
    margin: ${p1 =>
      css`
        ${p1.mv}px ${p2 => `${p2.mh}px`};
      `};
  `;
  expect(getStyle(Comp, { mv: 10, mh: 20 })).toMatchSnapshot();
});

test("default props", () => {
  const Comp = styled.div`
    display: block;
    margin: ${p1 =>
      css`
        ${p1.mv}px ${p2 => `${p2.mh}px`};
      `};
  `;

  Comp.defaultProps = {
    mv: 10,
    mh: 20
  };
  expect(getStyle(Comp)).toMatchSnapshot();
});

test("error", () => {
  expect(getStyle).toThrow();
});
