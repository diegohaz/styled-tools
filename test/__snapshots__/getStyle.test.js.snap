// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`default props 1`] = `
"
    display: block;
    margin: 
        10px 20px;
      ;
  "
`;

exports[`no props 1`] = `
"
    display: block;
    margin: 10px 20px;
  "
`;

exports[`with props 1`] = `
"
    display: block;
    margin: 
        10px 20px;
      ;
  "
`;
